rfcs/45-stageless.md

```diff
@@ -0,0 +1,984 @@
+We think that's good enough for now.
+
+(Maybe instead of `apply_system_buffers` applying the pending commands of *all* systems in the current schedule, we could construct `apply_system_buffers` from a tuple of system (set) labels, such that only those have their commands applied.
+It's debatable whether that would increase or reduce cognitive burden.
+That can be revisited later too.)
+
+### Why did you change states from a stack to a basic FSM? Where is `OnUpdate(X)`?
+
+If this RFC is merged and this design is implemented, we expect most of the problems that encouraged it to vanish.
+So rather than port pieces of the existing API, we think it's better to start fresh with a basic state machine and see what patterns emerge and what their limits are.
+If, after migration, a significant number of users are still turning towards a plugin to recover the lost stack functionality, we can consider upstreaming it again.
+
+A stack-based state machine should be trivial to implement though (i.e. with a loop inside the exclusive system).
+
+As for `OnUpdate(X)`... its interaction with `OnEnter(X)` was non-intuitive.
+Users often just wanted systems to run in their designated spots, but only when a certain state was active. That is much simpler to do now (see example below), so having a system set for this that you can only put in one spot seemed pointless.
+
+```rust
+fn main() {
+    App::new()
+        /* ... */
+        .add_system(
+            my_system
+            // A state being active has no inherent position in the schedule.
+            // Any system anywhere can be told to only run if a certain state is active.
+            .in_set(CoreSet::PostUpdate)
+            // This helper generates a condition that returns `true` if the state is currently `GameState::Playing`.
+            .run_if(state_equals(GameState::Playing))
+        )
+        /* ... */
+        .run();
+    )
+}
+```
+
+## Unresolved questions
+
+### What is the best way to approach migration?
+
+Whatever the changes, migration will be a big challenge.
+The proposed changes would affect the `App` and `Plugin` methods that all engine plugins and examples use, which will be a lot to review.
+At the same time, a large number of users are eagerly awaiting changes of this nature making it into the engine.
+
+If this RFC is merged, to expedite upstreaming while easing the burden on Bevy maintainers, we propose the following migration path:
+
+- Add all the new stuff in a new module (i.e. `bevy_ecs::stageless`) that will live alongside an unchanged `bevy_ecs::schedule`, unused (and not in the prelude).
+- Implement a trait (i.e. `StagelessAppExt`) for `App` that can hook into the new module. Ultimately temporary.
+- Upstream these to `main`.
+- Publish a migration guide for the new API.
+- Slowly port over our engine plugins to the new methods provided through `StagelessAppExt`.

```

Comment:
Slowly porting over seems nice if we can swing it. But how will this "dual mode" work in practice? How will we reconcile having some systems / plugins in the new schedule and some in the old schedule (and handling orderings correctly across schedules)? This "mixed schedule" approach seems like its own reasonably hard engineering problem that could introduce bugs specific to it.

Shouldn't the new schedule api be able to accommodate the current patterns? The initial porting should just be a matter of porting the core stages to the new set patterns, adding apply_buffers between each stage, renaming a few builder apis, and swapping out stage membership for set membership / orderings relative to specific apply_buffers instances (ex: FlushPostUpdate).

This seems easier and less error prone than trying to let both schedules and apis exist at the same time.

It seems like the "hard" part is adjusting the built in plugins to take advantage of the additional flexibility (and maybe blurring the lines a bit between core stages). This could / should be done piece by piece after the initial pr.

User migrations could follow a similar path. First do the quick and easy 1:1 "rephrase a few things to match the new apis". Easily covered in a migration guide. Then re-evaluate system orderings / relationships.

Reply:
> But how will this "dual mode" work in practice? How will we reconcile having some systems / plugins in the new schedule and some in the old schedule (and handling orderings correctly across schedules)?

My vision here is that we implement the machinery in its own module in successive PRs (`bevy_schedule2` or whatever).

Then, we port *all* of the app logic in a single massive but uncontroversial PR. Then once that's done, we remove the old API completely.

The goal is primarily to make the process easier for reviewers by focusing attention where it's needed.

Reply:
> But how will this "dual mode" work in practice?

I would expect plugins/apps to exist in a state of using one or other. The only overlap I can see is that plugins/apps using the old API could use the "run a schedule inside an exclusive system" pattern because `Systems` would be present. 

Nothing in the old API is broken by the new API and things added through the old API and new API would have no connection to each other.

> Shouldn't the new schedule api be able to accommodate the current patterns? 

Yes, we can replicate existing patterns in the new API.

---

Hopefully these wouldn't need to coexist for very long, but like Alice says, it's just a lot to review. Implementing the new behavior, updating all the plugins, and updating all the examples in a single PR will be extremely unpopular.

Splitting the core implementation work and migrating the internal plugins seemed like a reasonable choice.

Reply:
I don't think how feasible it is, but an alternative approach would be to have both the old and the new APIs in a usable state. That means that core plugins must have a double implementation, one for the current schedule and one for stageless.

That way, core plugins can be ported bit by bit to stageless by multiple contributors. In the meantime, the same can be done for the examples after the plugins they depend on get ported. I think this is a great opportunity for new contributors to participate in the project and familiarize both with the new API, and with how the various core plugins are organized. A project board to organize the work among the contributors may be ideal.

After everything is ported, the old API is removed.

One obvious risk of this approach is to have  functionality desync between pre-stageless and stageless plugins. This may also be less relevant than we think, since the bugs arise once an app/example transitions to stageless, and bugs are hopefully identified and fixed.

Another possible annoyance is that community plugin authors may have to do the double support as well, otherwise one app author may be forced to use the new API. This won't be a problem if the entire transition happens before 0.9 is released, but I doubt we can pull it off before the deadline without rushing.

Reply:
@Nilirad I believe that is the solution depicted in this RFC, to have two APIs in a working state, then gradually obsolete the old one until the point where the old one is completely replaced by the new one. In the last stage, given some time, most of the most important plugins should have been migrated. The ones that didn't will simply not be compatible from that version of Bevy on.

Reply:
> My vision here is that we implement the machinery in its own module in successive PRs (bevy_schedule2 or whatever).
> Then, we port all of the app logic in a single massive but uncontroversial PR. Then once that's done, we remove the old API completely.

This is also what I was going to push for. I see no real value in supporting a state where half of bevy plugins use the old system and half use the new system. And I see a lot of risk and complexity in trying to do it that way. (Which is how I interpreted the migration strategy in this RFC). Like you just said, I think we should develop the new schedule in parallel (in a `schedule2` module or something), but we should do the porting all at once (in the same step that we remove the old schedule and slot in the new one).

On the "bevy user" side, 99% of the breakage will just be a matter of renaming function calls. Imo its better to just rip the bandaid off and provide a good migration guide than create a bunch of mixed-mode machinery / let people do things "both ways".

Reply:
If porting all Bevy plugins and examples in a single migration is OK, then I'm all for it. I only proposed an incremental porting strategy because going the one big PR route was unpopular when I was headed that way.

In any case, I think there's low risk if for whatever reason we had to split it into a series of PRs.

Reply:
> If porting all Bevy plugins and examples in one go is OK, then I'm all for it.

With the currently proposed APIs, I think this is totally ok.

>  I only proposed an incremental porting strategy because going the one big PR route was unpopular when I was headed that way.

Yeah I think as an initial gut reaction to a large API rework, trying to avoid porting a large code base in one go is good. And I do think both approaches are valid / they both would work. But within the context of this change set, I think "one PR adding the new scheduling logic" and "one PR swapping it in and porting everything over" is our best bet.